File: neetcode/blind75/1D_DP/longest_increasing_subsequence.py
# https://neetcode.io/problems/longest-increasing-subsequence

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:

        dp = [1] * len(nums)

        # for starting index
        for i in range(len(nums) - 1, -1, -1):
            # for each element before, if less than starting value, it is part of a solution so in the dp index +1 to include this element in that solution
            for j in range(i, -1, -1):
                if (nums[j] < nums[i]):
                    dp[j] = max(dp[i] + 1, dp[j]) # must check max because it could be in the optimal solution

        return max(dp)


        # naive method, check every combination with a starting index and subsequence
        # max_sub = 0
        # def dfs(idx, curr_max, curr_sub):
        #     nonlocal max_sub
        #     if (idx >= len(nums)):
        #         return 0

        #     for i in range(idx, len(nums)):
        #         if (nums[i] > curr_max):
        #             dfs(i + 1, nums[i], curr_sub + 1)
        #             max_sub = max(max_sub, curr_sub + 1)

        # dfs(0, float('-inf'), 0)
        # return max_sub

File: neetcode/neet150_JS/1-D-Dynamic-Programming/longest-increasing-subsequence.js
// https://neetcode.io/problems/longest-increasing-subsequence

class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    lengthOfLIS(nums) {
        if (nums.length === 0) {
            return 0
        }

        const dp = Array(nums.length).fill(1)

        let max = 1

        for (let i = 0; i < nums.length; i ++) {
            for (let j = i; j < nums.length; j ++) {
                if (nums[j] > nums[i]) {
                    dp[j] = Math.max(dp[j], dp[i] + 1)

                    max = Math.max(max, dp[j])
                }
            }
        }
        return max
    }
}
